File: bindings/node/tests/completion_poller.ts
// SPDX-License-Identifier: MPL-2.0

const zlink = require('@zlink-systems/zlink');

/** Caller-owned public completion poller used by async binding contract tests. */
export class CompletionPollerDriver {
  private readonly poller = zlink.createPoller();
  private readonly events: any;
  private closed = false;

  constructor(sockets: any | readonly any[]) {
    const sources = Array.isArray(sockets) ? sockets : [sockets];
    this.events = zlink.createPollEvents(Math.max(1, sources.length));
    try {
      sources.forEach((socket, index) => {
        this.poller.add(socket, [zlink.PollEventFlag.PollCompletion], index);
      });
    } catch (error) {
      this.close();
      throw error;
    }
  }

  wait(timeoutMs = 5_000): number {
    return this.poller.wait(this.events, timeoutMs);
  }

  async settle<T>(promise: Promise<T>, timeoutMs = 5_000): Promise<T> {
    type Outcome = { readonly ok: true; readonly value: T }
      | { readonly ok: false; readonly error: unknown };
    let outcome: Outcome | undefined;
    void promise.then(
      value => { outcome = { ok: true, value }; },
      error => { outcome = { ok: false, error }; }
    );
    const deadline = Date.now() + timeoutMs;
    while (outcome === undefined) {
      const remaining = deadline - Date.now();
      if (remaining <= 0) throw new Error('completion poller deadline expired');
      this.wait(Math.min(remaining, 100));
      await Promise.resolve();
    }
    const completed = outcome as Outcome;
    if ('error' in completed) throw completed.error;
    return completed.value;
  }

  close(): void {
    if (this.closed) return;
    this.closed = true;
    this.events.close();
    this.poller.close();
  }
}

File: bindings/python/tests/completion_poller.py
import threading

import zlink


class CompletionPollerDriver:
    """Drive one public completion owner for async contract tests."""

    def __init__(self, socket):
        self._poller = zlink.create_poller()
        self._events = zlink.create_poll_events(1)
        self._stop = threading.Event()
        self._failure = None
        self._poller.add_socket(
            socket, zlink.PollEventFlag.POLLCOMPLETION, 1
        )
        self._thread = threading.Thread(
            target=self._run,
            name="zlink-test-completion-poller",
        )

    def _run(self):
        try:
            while not self._stop.is_set():
                self._poller.wait(self._events, 50)
        except BaseException as failure:
            self._failure = failure

    def __enter__(self):
        self._thread.start()
        return self

    def __exit__(self, exc_type, exc, tb):
        self._stop.set()
        self._thread.join(2)
        try:
            if self._thread.is_alive():
                raise AssertionError("completion poller thread did not stop")
            if self._failure is not None and exc_type is None:
                raise self._failure
        finally:
            self._poller.close()
